python/pyarrow/parquet.py

```diff
@@ -807,6 +807,80 @@ def write_table(table, where, row_group_size=None, version='1.0',
         writer.close()
 
 
+def write_to_dataset(table, where, partition_cols, root_path,
+                     preserve_index=True, **kwargs):
+    """
+    Wrapper around parquet.write_table for writing a Table to
+    Parquet format by partitions.
+    For each combination of partition columns and values,
+    a subdirectories are created in the following
+    manner:
+
+    root_dir/
+    ├── group1=value1
+    │   ├── group2=value1
+    │   │   └── <result.parquet>
+    │   └── group2=value2
+    │       └── <result.parquet>
+    ├── group1=value2
+    │   └── group2=value1
+    │       └── <result.parquet>
+    ...
+    └── group1=valueN
+        ├── group2=value1
+        │   └── <result.parquet>
+        ...
+        └── group2=valueN
+            └──<result.parquet>
+
+    Parameters
+    ----------
+    table : pyarrow.Table
+    where: string,
+        Name of the parquet file for data saved in each partition
+    parition_cols : list,
```

Comment:
Typo in parameter name.

This should be optional, so that this function also works for an unpartitioned dataset

Reply:
re Typo: oof!

re Optional parameter: if no parameter is specified, would it be best to just call the original `parquet.write_table`?

Reply:
No, because you're appending data to a dataset consisting of one or more Parquet files, so you might have:

```
root_dir/
  file1.parquet
  file2.parquet
  file3.parquet
  file4.parquet
```

So if `partition_cols` is not specified, then it will write another file to the folder. This special case should be handled in the implementation of this function to avoid unnecessary data copying, and unit tested, too

Reply:
Oh I see! Giving the file names UUIDs (as suggested above) will make this simpler and follow the spark implementation more closely: `root_dir` is specified, but each of the files appended under it will be `<uuid>.parquet`.

Reply:
re "Unnecessary data copying": Are you saying I should check if the data being appended to the dataset has duplicates within the existing dataset?  If so, this seems like it could get hairy: a) I might want duplicates in my dataset (not sure why I would) b) I'm not sure how to do it other than reading back in the existing data, appending new data, deduping, then overwriting the directory with the complete dataset. 

 I'm hoping users will know not to accidentally write the same data twice. Right now, my implementation for this is:
```
    if partition_cols is not None:
          (modified original function)
    else:
        outfile = compat.guid() + ".parquet"
        full_path = "/".join([root_path, outfile])
        write_table(table, full_path, **kwargs)
```